cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,532 @@
+
+```JavaScript
+var hpp = require('hpp');
+app.use(hpp());
+```
+
+### Return sanitized user objects
+
+Information about the users of an application is among the most critical information about the application. User tables generally include fields like id, username, full name, email address, birth date, password and in some cases social security numbers. Therefore, when querying and using user objects, you need to return only needed fields as it may be vulnerable to personal information disclosure. This is also correct for other objects stored on the database. If you just need a certain field of an object, you should only return the specific fields required. As an example, you can use a function like the following whenever you need to get information on a user. By doing so, you can only return the fields that are needed for your specific operation. In other words, if you only need to list names of the users available, you are not returning their email addresses or credit card numbers in addition to their full names.
+
+```JavaScript
+exports.sanitizeUser = function(user) {
+  return {
+    id: user.id,
+    username: user.username,
+    fullName: user.fullName
+  };
+};
+```
+
+### Use object property descriptors
+
+Object properties include 3 hidden attributes: `writable` (if false, property value cannot be changed), `enumerable` (if false, property cannot be used in for loops) and `configurable` (if false, property cannot be deleted). When defining an object property through assignment, these three hidden attributes are set to true by default. These properties can be set as follows:
+
+```JavaScript
+var o = {};
+Object.defineProperty(o, "a", {
+    writable: true,
+    enumerable: true,
+    configurable: true,
+    value: "A"
+});
+```
+
+Apart from these, there are some special functions for object attributes. `Object.preventExtensions()` prevents new properties from being added to the object.
+
+### Use access control lists
+
+Authorization prevents users from acting outside of their intended permissions. In order to do so, users and their roles should be determined with consideration of the principle of least privilege. Each user role should only have access to the resources they must use. For your Node.js applications, you can use [acl](https://www.npmjs.com/package/acl) module to provide ACL (access control list) implementation. With this module, you can create roles and assign users to these roles.
+
+## Error & Exception Handling
```

Comment:
We have https://github.com/OWASP/CheatSheetSeries/blob/33253edaa2218511193f91ad3da939b2d307f954/cheatsheets/Error_Handling_Cheat_Sheet.md maybe it should go there and here we will link to it? Any thoughts on that? @rbsec @ThunderSon @mleblebici

Reply:
I think keeping it here (maybe with a link to the error handling CS) is probably the best option for now.

That CS looks a bit dated, and I think there needs to be a wider discussion about how this sort of thing is handled in the language-specific cheat sheets. Easy enough to move it between them later if we decide to do that.